Updated Live Share capabilities page with LiveState and additional user scenarios. (#7959)
MicrosoftDocs/msteams-docs: msteams-platform/apps-in-teams-meetings/teams-live-share-capabilities.md

## Fluid distributed data structures

The Live Share SDK supports any [distributed data structure](https://fluidframework.com/docs/data-structures/overview/) included in Fluid Framework. These features serve as a set of primitives you can use to build robust collaborative scenarios, such as real-time updates of a task list or co-authoring text within an HTML `<textarea>`.

> Here's a quick overview of a few of the different types of objects available:

Reviewer: ```suggestion
Following are the different types of objects available:
```

## Live Share data structures

> The Live Share SDK includes a set of new Live Share `SharedObject` classes, which provide stateful and stateless objects. Unlike Fluid data structures, these features do not write changes to the Fluid container, enabling faster synchronization. These objects were designed from the ground up for common meeting scenarios for Teams meetings. Common scenarios include synchronizing what content the presenter is viewing, displaying metadata for each user in the meeting, or displaying a countdown timer.

Reviewer: ```suggestion
The Live Share SDK includes a set of new Live Share `SharedObject` classes, which provide stateful and stateless objects. Unlike Fluid data structures, `SharedObject` classes don’t write changes to the Fluid container, enabling faster synchronization. `SharedObject` classes were designed from the ground up for common meeting scenarios in Teams meetings. Common scenarios include synchronizing what content the presenter is viewing, displaying metadata for each user in the meeting, or displaying a countdown timer.
```

Author: `SharedObject` is a Fluid class and not something implemented by the Live Share SDK. We do extend it but in the context of this sentence, this is not accurate. I will update before committing.

| `LiveEvent`    | Broadcast individual events with any custom data attributes in the payload.                                                             |
| `LiveState`    | Similar to SharedMap, a distributed key-value store that allows for restricted state changes based on role, for example, the presenter. |
| `LiveTimer`    | Synchronize a countdown timer for a given interval.                                                                                     |

### LivePresence example

:::image type="content" source="../assets/images/teams-live-share/live-share-presence.png" alt-text="Screenshot shows an example of showing people who available in a sessionTeams using Live Share presence.":::

The `LivePresence` class makes tracking who is in the session easier than ever. When calling the `.initialize()` or `.updatePresence()` methods, you can assign custom metadata for that user, such as name, profile picture, or the identifier for content they are viewing. By listening to `presenceChanged` events, each client receives the latest `LivePresenceUser` object, collapsing all presence updates into a single record for each unique `userId`.

> Here are a few examples in which `LivePresence` can be used in your application:

Reviewer: ```suggestion
You can use ‘LivePresence' in your application in the following scenarios:
```

Author: These are just examples of scenarios and not an exhaustive list. Your wording implies that the provided scenarios are the only possible scenarios.

:::image type="content" source="../assets/images/teams-live-share/live-share-timer.png" alt-text="Screenshot shows an example of a count down timer with 9 seconds remaining.":::

> `LiveTimer` provides a simple countdown timer that is synchronized for everyone in a meeting. It is useful for scenarios that have a time limit, such as a group meditation timer or a round timer for a game.

Reviewer: ```suggestion
`LiveTimer` provides a simple countdown timer that is synchronized for everyone in a meeting. It’s useful for scenarios that have a time limit, such as a group meditation timer or a round timer for a game.
```

---

### LiveState example

:::image type="content" source="../assets/images/teams-live-share/live-share-state.png" alt-text="Screenshot shows an example of Live Share state to synchronize what planet in the solar system is actively presented to the meeting.":::

> The `LiveState` class enables synchronizing simple application state for everyone in a meeting. `LiveState` synchronizes two values: a `state` string and and a corresponding `data` object. This allows you to build an ephemeral distributed-state machine.

Reviewer: ```suggestion
The `LiveState` class enables synchronizing simple application state for everyone in a meeting. `LiveState` synchronizes two values: a `state` string and a corresponding `data` object, which allows you to build an ephemeral distributed-state machine.
```

> Here are a few examples in which `LiveState` can be used in your application:

Reviewer: ```suggestion
You can use ‘LiveState’ in your application in the following scenarios:
```

Author: These are just examples of scenarios and not an exhaustive list. Your wording implies that the provided scenarios are the only possible scenarios. We should not imply to developers that it can only be used for the scenarios described in this list.

> - Tracking the user identifier of the current presenter to build a "take control" feature.
> - Maintaining the content identifier that the current presenter is viewing (e.g., `taskId` on a task board).
> - Synchronizing the current step in a multi-round group activity (e.g., guessing phase in Agile Poker).

Reviewer: ```suggestion
- Tracking the user identifier of the current presenter to build a **take control** feature.
- Maintaining the content identifier that the current presenter is viewing. For example, `taskId` on a task board.
- Synchronizing the current step in a multi-round group activity. For example, guessing phase in the Agile Poker game.
```

> > [!NOTE]
> > Unlike `SharedMap`, once all users disconnect from a session the `state` and `data` values in `LiveState` will be reset.

Reviewer: ```suggestion
> [!NOTE]
> Unlike `SharedMap`, the `state` and `data` values in `LiveState` will be reset after all the users disconnect from a session.

```

  JUPITER = "Jupiter",
  SATURN = "Saturn",
  URANUS = "Uranus",
  NEPTUNE = "Neptune",
}

// Declare interface for type of custom data for user
interface ICustomState {
  planetName: PlanetName;
}

// Join the Fluid container
const host = LiveShareHost.create();
const liveShare = new LiveShareClient(host);
const schema = {
  initialObjects: {
    appState: LiveState<ICustomState>,
  },
};
const { container } = await liveShare.joinContainer(schema);
const appState = container.initialObjects.appState as LiveState<ICustomState>;

// Register listener for changes to state and corresponding custom data
appState.on("stateChanged", (state: string, data: ICustomState | undefined, local: boolean) => {
  if (state === "planet-viewer") {
    const planetName = data?.planetName;
    // Update app to display an image of the selected planet for the selected solar system
  } else {
    // No planet is yet selected
  }
});

// Set roles who can change state and start listening for changes
appState.initialize();

function onSelectPlanet(planetName: PlanetName) {
  appState.changeState("planet-viewer", {
    planetName,
  });
}
```

---

## Role verification for live data structures

> Meetings in Teams include calls, all-hands meetings, and online classrooms. Meeting participants may span across organizations, have different privileges, or simply have different goals. Thus, it is important to respect the privileges of different user roles during meetings. Live objects are designed to support role verification, allowing you to define the roles that are allowed to send messages for each individual live object. For example, you could choose that only meeting presenters and organizers can control video playback, but still allow guests and attendees to request videos to watch next.

Reviewer: ```suggestion
Meetings in Teams include calls, all-hands meetings, and online classrooms. Meeting participants might span across organizations, have different privileges, or simply have different goals. Hence, it’s important to respect the privileges of different user roles during meetings. Live objects are designed to support role verification, allowing you to define the roles that are allowed to send messages for each individual live object. For example, you could choose that only meeting presenters and organizers can control video playback, but still allow guests and attendees to request videos to watch next.
```

> [!NOTE]
> The `LivePresence` class doesn't support role verification. The `LivePresenceUser` object has a `getRoles` method, which returns the meeting roles for a given user.

> Here is a simple example using `LiveState` for to synchronize which user is the active presenter, where only presenters and organizers can take control:

Reviewer: ```suggestion
In the following example where only presenters and organizers can take control, `LiveState` is used to synchronize which user is the active presenter. 
```
